XMonad/Doc/Extending.hs

```diff
@@ -236,111 +230,148 @@ yourself.
-provide useful functions for editing your key bindings; "XMonad.Util.EZConfig" even lets you use emacs-style keybinding descriptions like \"M-C-<F12>\".
+> main :: IO ()
+> main = xmonad $ def { keys = myKeys }
+>  where
+>   myKeys :: XConfig l -> Map (ButtonMask, KeySym) (X ())
+>   myKeys _conf = Map.fromList
+>     [ ((mod1Mask, xK_m        ), spawn "echo 'Hi, mom!' | dzen2 -p 4")
+>     , ((mod1Mask, xK_BackSpace), spawn "xterm")
+>     ]
+
+Now, obviously we don't want to do that; we only want to add to existing
+bindings (or, perhaps, override some of them with our own).  Let's break
+@myKeys@ down a little.  You can think of the type signature like this:
+
+>    myKeys :: UserConfig -> Map KeyPress Action
+
+It takes some user config and, from that, produces a map that associates
+certain keypresses with actions to execute.  Because we're setting @keys
+= myKeys@ above, this is also the type of 'XMonad.Core.keys'.
+
+This means that, as a first guess, the type signature of our version of
+'XMonad.Util.EZConfig.additionalKeys' might look like
+
+> myAdditionalKeys :: XConfig l
+>                     -- ^ Base config with xmonad's default keybindings
+>                  -> (XConfig l -> Map (ButtonMask, KeySym) (X ()))
+>                     -- ^ User supplied keybindings
+>                  -> XConfig l
+>                     -- ^ Resulting config with everything merged together
+
+However, even assuming a correct implementation, using this is not very
+ergonomic:
+
+> main = xmonad $ def
+>  `myAdditionalKeys`
+>    (\ _conf -> Map.fromList
+>      [ ((mod1Mask, xK_m        ), spawn "echo 'Hi, mom!' | dzen2 -p 4")
+>      , ((mod1Mask, xK_BackSpace), spawn "xterm")
+>      ])
+
+Having to specify a lambda with parentheses and call
+'Data.Map.Strict.fromList' does not make for a good user experience.
+Since one /always/ has to call that function anyways, we may well just
+accept a list from the user and transform it to a map ourselves.  As an
+additional simplification, how about we don't care about the config
+argument at all and simply ask the user for a list?  The resulting
+signature
+
+> myAdditionalKeys :: XConfig l
+>                  -> [(ButtonMask, KeySym), (X ())]
+>                  -> XConfig l
+
+looks exactly like what we want!  Now that we know what kind of data
+structure—that is, maps—we are dealing with, the implementation of this
+function simply merges the two together, preferring the user config to
+xmonad's defaults in case of any conflicts.  Thankfully, someone else
+has already done the hard work and written the merging function for us;
+it's called 'Data.Map.Strict.union'.
+
+What's left is essentially playing type tetris:
+
+> myAdditionalKeys baseConf keyList =
+>   let mergeKeylist conf = Map.fromList keyList `Map.union` (keys baseConf) conf
+>    in baseConf { keys = mergeKeylist }
+
+The function @mergeKeyList@ take some user config, transforms the custom
```

Comment:
But why does it take a user config at all? I think this question should be answered as soon as the `myKeys :: UserConfig -> Map KeyPress Action` example. I mean, I know the answer, yet I'm still scratching my head here! :-)

Reply:
`f conf = … (modMask conf) …` is shorter but not sure if good for what you want to say. No strong opinions here, I'll leave it up to you to decide which sounds better.

Reply:
It is shorter indeed, though I like the fact that `XConfig{modMask=modMask}` makes it quite clear we are _only_ interested in that field (I just realized that we can even drop the `conf@` in these examples)).  It probably does not matter much in the end.  The core message that it would be nice if the keybindings knew things about the config will probably come across either way

Reply:
c8417409fcdd6ef3dd02a1d92e7b17edd13b0fc6?